### PR TITLE
T6 animtree decompression compression

### PR DESCRIPTION
**Pull request by JezuzLizard**
Add the ability to automatically decompress animtrees when dumping, and compress them when linking for T6.

### CODE REVIEW

File: src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp [hunk 1]
```diff
@@ -20,13 +26,61 @@ bool AssetLoaderRawFile::CanLoadFromRaw() const
...
+    if (ret != Z_OK)
+    {
+        throw std::runtime_error("Initializing deflate failed");
+    }
+
+    ret = deflate(&zs, Z_FINISH);
+
+    if (ret != Z_STREAM_END)
+    {
+        std::cout << "Deflate failed for loading animtree file \"" << assetName << "\"" << std::endl;
```
Comment (diamante0018) [inline src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp:62]: please consider using std::cerr for printing error messages (here and elsewhere in the pr)

**Comment by diamante0018** [comment]
you need to install the llvm-project on your system and format your code with clang-format

https://clang.llvm.org/docs/ClangFormat.html

File: src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp [hunk 2]
```diff
@@ -1,12 +1,88 @@
...
+    size_t writtenSize = 0;
+    while (zs.avail_in > 0)
+    {
+        zs.next_out = buffer;
+        zs.avail_out = sizeof buffer;
+        ret = inflate(&zs, Z_SYNC_FLUSH);
+
+        if (ret < 0)
+        {
+            std::cout << "Inflate failed for dumping animtree file \"" << rawFile->name << "\"" << std::endl;
```
Comment (Laupetin) [inline src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp:64]: Same would apply here -> std::cerr

File: src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp [hunk 3]
```diff
@@ -1,12 +1,88 @@
...
+        if (writtenSize + inflateOutSize >= outLen)
+        {
+            // Last byte is a \0 byte. Skip it.
+            stream.write(reinterpret_cast<char*>(buffer), inflateOutSize - 1);
+        }
+        else
+        {
+            stream.write(reinterpret_cast<char*>(buffer), inflateOutSize);
+        }
+        writtenSize += inflateOutSize;
```
Comment (Laupetin) [inline src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp:80]:
```suggestion
        const auto inflateOutSize = sizeof buffer - zs.avail_out;
        stream.write(reinterpret_cast<char*>(buffer), inflateOutSize);
        writtenSize += inflateOutSize;
```

I assume you used the code from T5 here. While the last byte being an unnecessary zero byte applies there, it does not seem to be the case for T6 animtrees.

Therefore currently the last character seems to be cut off.

**Review by Laupetin — COMMENTED** [review]
Thank you, good addition :)
Just a remark that T6 seems to have no zero byte at the end unlike T5.
Otherwise looks good 👍

File: src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp (hunk 1 above)
Comment (JezuzLizard) [reply src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp]: Fixed.

File: src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp (hunk 2 above)
Comment (JezuzLizard) [reply src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp]: Done.

File: src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp (hunk 3 above)
Comment (JezuzLizard) [reply src/ObjWriting/Game/T6/AssetDumpers/AssetDumperRawFile.cpp]: Done. Also removed the unused variable writtenSize.

File: src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp [hunk 4]
```diff
@@ -20,13 +26,61 @@ bool AssetLoaderRawFile::CanLoadFromRaw() const
...
 {
-    const auto file = searchPath->Open(assetName);
-    if (!file.IsOpen())
+    const auto uncompressedBuffer = std::make_unique<char[]>(static_cast<size_t>(file.m_length + 1));
+    file.m_stream->read(uncompressedBuffer.get(), file.m_length);
+    if (file.m_stream->gcount() != file.m_length)
+        return false;
+    uncompressedBuffer[static_cast<size_t>(file.m_length)] = '\0';
+
+    const auto compressionBufferSize = static_cast<size_t>(file.m_length + 1 + sizeof(uint32_t) + COMPRESSED_BUFFER_SIZE_PADDING);
```
[the remark this answers is not in the document]
Comment (JezuzLizard) [reply src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp]: Done.

File: src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp [hunk 5]
```diff
@@ -20,13 +26,61 @@ bool AssetLoaderRawFile::CanLoadFromRaw() const
...
+    auto* compressedBuffer = static_cast<char*>(memory->Alloc(compressionBufferSize));
+
+    z_stream_s zs{};
+
+    zs.zalloc = Z_NULL;
+    zs.zfree = Z_NULL;
+    zs.opaque = Z_NULL;
+    zs.avail_in = static_cast<uInt>(file.m_length + 1);
```
[the remark this answers is not in the document]
Comment (JezuzLizard) [reply src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp]: Addressed.

File: src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp [hunk 6]
```diff
@@ -20,13 +26,61 @@ bool AssetLoaderRawFile::CanLoadFromRaw() const
...
+    if (ret != Z_STREAM_END)
+    {
+        std::cout << "Deflate failed for loading animtree file \"" << assetName << "\"" << std::endl;
+        deflateEnd(&zs);
         return false;
+    }
+
+    const auto compressedSize = compressionBufferSize + sizeof(uint32_t) - zs.avail_out;
+
+    reinterpret_cast<uint32_t*>(compressedBuffer)[0] = static_cast<uint32_t>(file.m_length + 1); // outLen
```
[the remark this answers is not in the document]
Comment (JezuzLizard) [reply src/ObjLoading/Game/T6/AssetLoaders/AssetLoaderRawFile.cpp]: Done.

**Comment by Laupetin** [comment]
Looks good :D

**Pull request merged**